(feat) O3-5027: Implement beautiful configurable login page

UjjawalPrabhat opened the pull request:
# Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. Ensure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing#contributing-guidelines) label (such as `feat`, `fix`, or `chore`, among others). See existing PR titles for inspiration.

## If applicable

- [x] My work is based on designs, which are linked or shown either in the Jira ticket or the description below.
- [x] My work includes tests or is validated by existing tests.
- [x] I have updated the [esm-framework mock](https://github.com/openmrs/openmrs-esm-core/blob/main/packages/framework/esm-framework/mock.tsx) to reflect any API changes I have made.

## Summary
This PR implements a beautiful and configurable login page for OpenMRS 3.x as part of the OMRS Hackathon 2025. The login page now supports:
- Configurable images and branding
- Custom messages for planned downtime
- Improved visual design and user experience
- Responsive layout for different screen sizes

## Related Issue
https://issues.openmrs.org/browse/O3-5027

## Other
This work was completed as part of the OpenMRS Hackathon 2025. The implementation focuses on making the login page easily configurable while maintaining a beautiful and professional appearance.

sherrif10 commented on packages/apps/esm-login-app/src/background/background-wrapper.component.tsx:
```diff
@@ -0,0 +1,74 @@
+
+      case 'gradient':
+        if (background.value) {
+          style.background = background.value;
```
Hello @UjjawalPrabhat , We need Loginbackground to be configurable, i went ahead and tested your code and it seems like something is missing. Check this out. 
[loginbackground.webm](https://github.com/user-attachments/assets/60486806-649c-4554-a189-fcd2f25e26bd)

UjjawalPrabhat replied:
Hey @sherrif10, the Logo change config you're referring to wasn't implemented by me. It was there from before and therefore the logo is editable.
Do you want me to make the Logo constant?
Also please let me know if there's any other changes or enhancements required.

sherrif10 replied:
We need the login page to look like this . Share the vedio of  your changes before and after 
<img width="1600" height="844" alt="unnamed" src="https://github.com/user-attachments/assets/064e479e-fb90-474d-b6ed-2afc5f3a56eb" />

UjjawalPrabhat replied:
@sherrif10 could you send the images used in this the login page?

sherrif10 replied:
Hello @UjjawalPrabhat , We want to come up with something like that in the image i shared, FYI , Login page is already configurable, the only changes needed is to make login page appear like that.

UjjawalPrabhat replied:
@sherrif10 I'm on it. I was talking about the images used like the 3 doctor's and logo image used in the login page. It'd be easier for me to configure then

sherrif10 replied:
Yes it needs to be configurable and when the configuration is set to default, Default OMRS login page  should appear, Then you should consider making it appear left or right on the  login page.

tendomart commented on packages/apps/esm-login-app/src/system-messages/system-messages.component.tsx:
```diff
@@ -0,0 +1,88 @@
+import React, { useState, useCallback, useMemo } from 'react';
```
I agree with @sherrif10  we should first focus on having an image side by side with the login form he shared above

hey @UjjawalPrabhat  did you have a chance to look at how it was implemented at https://github.com/METS-Programme/esm-ugandaemr-login ?
 I don't think you really need to come up with something  completely new

UjjawalPrabhat replied:
> I agree with @sherrif10 we should first focus on having an image side by side with the login form he shared above
> 
> hey @UjjawalPrabhat did you have a chance to look at how it was implemented at https://github.com/METS-Programme/esm-ugandaemr-login ? I don't think you really need to come up with something completely new

@tendomart Thanks for this. I actually implemented it from scratch and then saw this when i was about to commit. But I made all the required changes including login page switch b/w uganda-emr and default, configurable system-messages and background.
@sherrif10 Please review my changes and let me know if any further changes required.

sherrif10 commented on packages/apps/esm-login-app/src/background/background-wrapper.component.tsx:
```diff
@@ -0,0 +1,137 @@
+import styles from './background-wrapper.scss';
+
+// Import Uganda EMR background for default usage
+const getUgandaEmrBackgroundUrl = () => {
```
Also remember for all this to happen out of the box. You need to map this image on the server either via dockerFile in openmrs distro to let docker pick the path such that immediately you start openmrs-esm-login -app, it can pick that Path and renders it via config that we made in esm-login-app. cc @reagan-meant  @ibacher

sherrif10 commented on packages/apps/esm-login-app/src/background/background-wrapper.scss:
```diff
@@ -0,0 +1,102 @@
+  }
+}
+
+// Uganda EMR specific background styles (from original app)
```
Please remove those openmrs images and ugandaEMR image. Those are supposed to be handled by the server.

UjjawalPrabhat committed:
refactor(login): address review feedback on PR scope

Significantly reduced the number of changed files based on @ibacher's
comments. Removed all framework modifications and simplified the
background image handling to use direct URLs.

All the configurable login features are still there - just implemented
more cleanly now. Should be much easier to review and merge.

UjjawalPrabhat commented:
@ibacher @sherrif10 @tendomart 

Fixed the scope issues! Went from 49 changed files down to 11, and they're all just in the login app now. 

Removed the framework stuff and simplified the image loading like you asked - now it just uses URLs directly instead of bundling assets.

All the login customization still works (backgrounds, layouts, branding, etc.) but it's much cleaner implementation.

Ready for another look when you get a chance 👍

sherrif10 commented on packages/apps/esm-login-app/src/login/login.scss:
```diff
@@ -156,3 +156,127 @@
   left: 0;
 }
+
+// Configurable branding styles
```
Remove this comments, Then feel free to share the latest vedio of this changea

UjjawalPrabhat replied:
@sherrif10 you want me to remove all comments throughout my code changes or just this one?

UjjawalPrabhat committed:
Merge branch 'main' into feature/configurable-login-page

jwnasambu committed:
Merge branch 'main' into feature/configurable-login-page

ibacher reviewed (commented):
A few notes almost entirely on the configuration. Note that the idea here is to allow some simple, light customization easily. We shouldn't have nearly so many CSS-specific properties exposed as configuration options.

Let's focus on primarily supporting single color backgrounds and background images and try to minimize additional properties. We can add them as and when they become relevant.

UjjawalPrabhat committed:
refactor(config): simplify login config schema

- Reduce schema from 386 to 194 lines (50% reduction)
- Add _type: Type.Object to all nested objects
- Rename columnPosition  loginFormPosition
- Remove complex CSS styling (card, button sections)
- Simplify background to color + imageUrl only
- Support translation keys in all text fields
- Remove gradient, overlay, contactEmail, customLinks
- All 38 tests passing

UjjawalPrabhat committed:
Merge branch 'main' into feature/configurable-login-page

UjjawalPrabhat committed:
Merge branch 'main' into feature/configurable-login-page

UjjawalPrabhat committed:
Merge branch 'main' into feature/configurable-login-page

UjjawalPrabhat committed:
Merge branch 'main' into feature/configurable-login-page

UjjawalPrabhat commented:
Hey @sherrif10 @ibacher @denniskigen @tendomart ,

Just checking in on this PR: [#1460](https://github.com/openmrs/openmrs-esm-core/pull/1460). The implementation now closely matches all requested feedback: 
- config schema is simplified, login background/image works via direct URL, and all changes are scoped only to the login app.
- All tests are passing and everything builds cleanly.

Would appreciate a final review or merge whenever you have time. Let me know if there’s anything else needed!

UjjawalPrabhat committed:
Merge branch 'main' into feature/configurable-login-page

UjjawalPrabhat committed:
Merge branch 'main' into feature/configurable-login-page

UjjawalPrabhat committed:
Merge branch 'main' into feature/configurable-login-page

UjjawalPrabhat committed:
Merge branch 'main' into feature/configurable-login-page

UjjawalPrabhat committed:
Merge branch 'main' into feature/configurable-login-page

UjjawalPrabhat committed:
Merge branch 'main' into feature/configurable-login-page

Yashcat2 reviewed (commented):
Left a few inline comments on the config-schema.ts mainly around validation and error handling.

ibacher commented:
@UjjawalPrabhat Is this continuing to be worked on? It's still a draft.
